src/samples/audio/smart_amp_test.c

```diff
@@ -665,6 +617,31 @@ static int smart_amp_copy(struct comp_dev *dev)
 					  &sink_buf->stream);
 
 	k_mutex_lock(&sad->lock, K_FOREVER);
+#if CONFIG_IPC_MAJOR_4
+	/* check for feedback source buffer as the feedback pipeline could have started later */
+	if (!sad->feedback_buf) {
+		struct list_item *blist;
+
+		list_for_item(blist, &dev->bsource_list) {
+			struct comp_buffer __sparse_cache *buffer_c;
+			struct comp_buffer *source_buffer;
+
+			source_buffer = container_of(blist, struct comp_buffer, sink_list);
+			buffer_c = buffer_acquire(source_buffer);
+			if (IPC4_SINK_QUEUE_ID(buffer_c->id) == SOF_SMART_AMP_FEEDBACK_QUEUE_ID) {
+				sad->feedback_buf = source_buffer;
+				buffer_c->stream.channels = sad->config.feedback_channels;
+				buffer_c->stream.rate =
+					sad->ipc4_cfg.base.audio_fmt.sampling_frequency;
+
+				buffer_release(buffer_c);
+				break;
+			}
+
+			buffer_release(buffer_c);
+		}
+	}
+#endif
```

Comment:
one risk that : smart_amp_copy and ipc4 unbind run in different threads. It is possible that playback is using the feedback buffer and capture is stopped and unbind then the buffer is removed. 

The original code use mutex to sync two threads.

Reply:
Kai suggested that we should add ref count to free buffer to avoid such issue, but now we just free buffer

Reply:
this is a valid point. The only way then would be to check at every copy(). I wonder if that's acceptable

Reply:
Can we use notifier method ? currently we have
````c
enum notify_id {
	NOTIFIER_ID_CPU_FREQ = 0,		/* struct clock_notify_data * */
	NOTIFIER_ID_SSP_FREQ,			/* struct clock_notify_data * */
	NOTIFIER_ID_KPB_CLIENT_EVT,		/* struct kpb_event_data * */
	NOTIFIER_ID_DMA_DOMAIN_CHANGE,		/* struct dma_chan_data * */
	NOTIFIER_ID_BUFFER_PRODUCE,		/* struct buffer_cb_transact* */
	NOTIFIER_ID_BUFFER_CONSUME,		/* struct buffer_cb_transact* */
	NOTIFIER_ID_BUFFER_FREE,		/* struct buffer_cb_free* */
	NOTIFIER_ID_DMA_COPY,			/* struct dma_cb_data* */
	NOTIFIER_ID_LL_POST_RUN,		/* NULL */
	NOTIFIER_ID_DMA_IRQ,			/* struct dma_chan_data * */
	NOTIFIER_ID_DAI_TRIGGER,		/* struct dai_group * */
	NOTIFIER_ID_COUNT
};
````

Reply:
I have modified to check for feedback buf in every copy. Its not that bad for the smart amp because it really only has 2 source buffers.

Reply:
yes, but still not address the multi-thread risk

Reply:
````c
diff --git a/src/include/sof/lib/notifier.h b/src/include/sof/lib/notifier.h
index 06762b071..39e17cfd4 100644
--- a/src/include/sof/lib/notifier.h
+++ b/src/include/sof/lib/notifier.h
@@ -34,6 +34,8 @@ enum notify_id {
        NOTIFIER_ID_LL_POST_RUN,                /* NULL */
        NOTIFIER_ID_DMA_IRQ,                    /* struct dma_chan_data * */
        NOTIFIER_ID_DAI_TRIGGER,                /* struct dai_group * */
+       NOTIFIER_ID_MODULE_BIND,                /* struct ipc4_module_bind_unbind */
+       NOTIFIER_ID_MODULE_UNBIND,              /* struct ipc4_module_bind_unbind */
        NOTIFIER_ID_COUNT
 };

diff --git a/src/ipc/ipc4/helper.c b/src/ipc/ipc4/helper.c
index 7ebe12449..5c6525af3 100644
--- a/src/ipc/ipc4/helper.c
+++ b/src/ipc/ipc4/helper.c
@@ -16,6 +16,7 @@
 #include <ipc/dai.h>
 #include <sof/ipc/msg.h>
 #include <sof/lib/mailbox.h>
+#include <sof/lib/notifier.h>
 #include <sof/list.h>
 #include <sof/platform.h>
 #include <rtos/wait.h>
@@ -375,6 +376,9 @@ int ipc_comp_connect(struct ipc *ipc, ipc_pipe_comp_connect *_connect)
        if (ret < 0)
                goto e_src_bind;

+       notifier_event(NULL, NOTIFIER_ID_MODULE_BIND,
+                      NOTIFIER_TARGET_CORE_ALL_MASK, bu, sizeof(*bu));
+
        ret = comp_buffer_connect(sink, sink->ipc_config.core, buffer,
                                  PPL_CONN_DIR_BUFFER_TO_COMP);
        if (ret < 0) {
@@ -386,6 +390,9 @@ int ipc_comp_connect(struct ipc *ipc, ipc_pipe_comp_connect *_connect)
        if (ret < 0)
                goto e_sink_bind;

+       notifier_event(NULL, NOTIFIER_ID_MODULE_BIND,
+                      NOTIFIER_TARGET_CORE_ALL_MASK, bu, sizeof(*bu));
+
        /* update direction for sink component if it is not set already */
        if (!sink->direction_set && source->direction_set) {
                sink->direction = source->direction;
@@ -407,10 +414,14 @@ int ipc_comp_connect(struct ipc *ipc, ipc_pipe_comp_connect *_connect)
        return IPC4_SUCCESS;
 err:
        comp_unbind(sink, bu);
+       notifier_event(NULL, NOTIFIER_ID_MODULE_UNBIND,
+                  NOTIFIER_TARGET_CORE_ALL_MASK, bu, sizeof(*bu));
 e_sink_bind:
        pipeline_disconnect(sink, buffer, PPL_CONN_DIR_BUFFER_TO_COMP);
 e_sink_connect:
        comp_unbind(source, bu);
+       notifier_event(NULL, NOTIFIER_ID_MODULE_UNBIND,
+                  NOTIFIER_TARGET_CORE_ALL_MASK, bu, sizeof(*bu));
 e_sink_bind:
        pipeline_disconnect(sink, buffer, PPL_CONN_DIR_BUFFER_TO_COMP);
 e_sink_connect:
        comp_unbind(source, bu);
+       notifier_event(NULL, NOTIFIER_ID_MODULE_UNBIND,
+          NOTIFIER_TARGET_CORE_ALL_MASK, bu, sizeof(*bu));
 e_src_bind:
        pipeline_disconnect(source, buffer, PPL_CONN_DIR_COMP_TO_BUFFER);
 free:
@@ -485,6 +496,12 @@ int ipc_comp_disconnect(struct ipc *ipc, ipc_pipe_comp_connect *_connect)
        if (sink->pipeline->status < COMP_STATE_ACTIVE)
                pipeline_disconnect(sink, buffer, PPL_CONN_DIR_BUFFER_TO_COMP);

+       notifier_event(NULL, NOTIFIER_ID_MODULE_UNBIND,
+          NOTIFIER_TARGET_CORE_ALL_MASK, bu, sizeof(*bu));
+       notifier_event(NULL, NOTIFIER_ID_MODULE_UNBIND,
+          NOTIFIER_TARGET_CORE_ALL_MASK, bu, sizeof(*bu));
+
+
        ret = comp_unbind(src, bu);
        ret1 = comp_unbind(sink, bu);
        irq_local_enable(flags);
diff --git a/src/samples/audio/smart_amp_test.c b/src/samples/audio/smart_amp_test.c
index 1b84a393d..ef759c53d 100644
--- a/src/samples/audio/smart_amp_test.c
+++ b/src/samples/audio/smart_amp_test.c
@@ -7,7 +7,8 @@
 #include <sof/compiler_attributes.h>
 #include <sof/samples/audio/smart_amp_test.h>
 #include <sof/audio/ipc-config.h>
-#include <sof/trace/trace.h>
+#include <sof/trace/trace.h>   
+#include <sof/lib/notifier.h>
 #include <sof/ipc/msg.h>
 #include <sof/ut.h>
@@ -44,6 +45,51 @@ struct smart_amp_data {
        uint32_t out_channels;
 };

+static void smart_amp_notifier_bind(void *arg, enum notify_id type, void *data)
+{
+       struct comp_dev *dev = arg;
+       struct smart_amp_data *sad = comp_get_drvdata(dev);
+       struct comp_buffer __sparse_cache *buffer_c;
+       struct comp_buffer *source_buffer;
+       struct list_item *blist;
+
+       comp_err(dev, "=========== smart_amp_notifier_bind()");
+
+       /* searching for feedback source buffers */
+       list_for_item(blist, &dev->bsource_list) {
+               source_buffer = container_of(blist, struct comp_buffer, sink_list);
+               buffer_c = buffer_acquire(source_buffer);
+               if (IPC4_SINK_QUEUE_ID(buffer_c->id) == SOF_SMART_AMP_FEEDBACK_QUEUE_ID) {
+                       k_mutex_lock(&sad->mutex, K_FOREVER);
+                       sad->feedback_buf = source_buffer;
+
+                       buffer_c->stream.channels = sad->config.feedback_channels;
+                       buffer_c->stream.rate = sad->config.base.audio_fmt.sampling_frequency;
+                       buffer_release(buffer_c);
+                       k_mutex_unlock(&sad->mutex);
+
+                       break;
+               }
+
+               buffer_release(buffer_c);
+       }
+}
+
+static void smart_amp_notifier_unbind(void *arg, enum notify_id type, void *data)
+{
+       struct comp_dev *dev = arg;
+       struct smart_amp_data *sad = comp_get_drvdata(dev);
+       struct ipc4_module_bind_unbind *bu = data;
+
+       comp_err(dev, "=========== smart_amp_notifier_unbind()");
+
+       if (bu->extension.r.dst_instance_id == SOF_SMART_AMP_FEEDBACK_QUEUE_ID) {
+               k_mutex_lock(&sad->mutex, K_FOREVER);
+               sad->feedback_buf = NULL;
+               k_mutex_unlock(&sad->mutex);
+       }
+}
 static struct comp_dev *smart_amp_new(const struct comp_driver *drv,
                                      const struct comp_ipc_config *config,
                                      const void *spec)
@@ -75,6 +121,12 @@ static struct comp_dev *smart_amp_new(const struct comp_driver *drv,
 #if CONFIG_IPC_MAJOR_4
        k_mutex_init(&sad->mutex);

+       notifier_register(dev, NULL, NOTIFIER_ID_MODULE_BIND,
+                         smart_amp_notifier_bind, 0);
+
+       notifier_register(dev, NULL, NOTIFIER_ID_MODULE_UNBIND,
+                         smart_amp_notifier_unbind, 0);
+
        cfg = spec;
        bs = sizeof(struct sof_smart_amp_config);
 #else

````

Reply:
I want to totally remove bind & unbind ops for ipc4 bind msg and use notifier to replace them

Reply:
> yes, but still not address the multi-thread risk

why not? it's not like an unbind will happen during a copy right?

Reply:
@RanderWang ok i will close this PR. Why dont you submit your version to remove bind/unbind?

Reply:
BTW how does this notifier work? lets say a comp bind/unbind happens, which is in the IPC context, will the notifier be scheduled as a new task to be executed after the IPC task is done? what I am asking is if the notifier callback will be guaranteed to execute before the next copy.

Reply:
> @RanderWang ok i will close this PR. Why dont you submit your version to remove bind/unbind?

I plan to but I am busy with multi core feature and AEC. It is in my OKR

Reply:
> BTW how does this notifier work? lets say a comp bind/unbind happens, which is in the IPC context, will the notifier be scheduled as a new task to be executed after the IPC task is done? what I am asking is if the notifier callback will be guaranteed to execute before the next copy.

Sure. There is also a multi-thread risk. We still need to add ref count to avoid issue.

Reply:
concerning the multithreading race: this is again largely a question of whether we allow running connected pipelines on different cores. So far we don't. Then, if both these events - copying and unbinding - are happening on the same core, what contexts they're running in and can they preempt each other? IIRC unbinding would be happening in an IPC thread on core 0 or in the IDC thread (p4wq work with Zephyr) on other cores. Pipeline threads have the highest priority, so it's a question of whether they can preempt IPC and IDC contexts

Reply:
is it for module adapter ?  You should add some comment in TODO.  We need add refcounting first to avoid multi-thread risk.

Reply:
> is it for module adapter ? You should add some comment in TODO. We need add refcounting first to avoid multi-thread risk.

@RanderWang could you please describe the scenario that you find problematic and that would need an addition of a reference counter?

Reply:
@lyakh  I meant removing bind & unbind ops in smart_test which we use mutex to sync with ipc thread and smart_copy thread.  Smart AMP saves a pointer of feedback buffer which is from another stream.  smart AMP use it for copy and a risk that ipc unbind will delete it since the feedback stream is stopped when it is used by smart AMP. We can't delete it since it is used by smart AMP. With reference counter we can address it.

Reply:
sad->feedback_buf is not synchronized with ipc unbind thread. The mutex is not used by ipc thread now and only used by smart thread, so it can't protect feedback_buf now.

Reply:
but i assign the feedback buf before processing and unset it right after in every copy. its not optimal but there's no risk with multithreading here

Reply:
yes, but it is not protected when it is using by copy unless using some mutex to protect feedback buffer to avoiding free or reused.

Reply:
in original code both smart amp and ipc thread use mutex to access feedback buffer, but now ipc thread doesn't use it

Reply:
@RanderWang could you sketch a sequence of events when you think protection will be required and without one data will be corrupted?

Reply:
Two usage of feedback buffer,  smart copy  and ipc bind/unbind message for capture stream.  We need to take care of feedback buffer is being used by smart copy when ipc unbind msg is processed to delete the feedback buffer for capture stream. I will add refer count for buffer to avoid competition

Reply:
@RanderWang sorry, maybe you could do something like https://github.com/thesofproject/sof/pull/7089#issuecomment-1433117586 - just list the sequence of steps / function calls that you think can cause a problem

Reply:
Two streams : smart amp playback and ssp capture
(1) set up two streams
(2) bind smart_amp.2.1 to copier.ssp.8.1, a feedback buffer will be created to connect two streams
(3) start playback stream one core 0, capture stream on core 1
(4) smart amp playback is copying data from feedback buffer
(5) capture stream is stopped and smart_amp.2.1 is unbind with copier.ssp.8.1, feedback buffer is disconnected and free
     at this time step (4) is still using the feedback buffer

![smart](https://user-images.githubusercontent.com/39945651/220637887-0aa8c896-93a9-4774-8e20-793e7cafb472.png)

Reply:
@RanderWang running connected pipelines on different cores is still unsupported.

Reply:
...which also incidentally means, that this is none of smart-amp specific issue, this would be the case with any connected pipelines on different cores - with copier, mixer or anything

Reply:
@lyakh  The major issue is that unbind will disconnect buffer and free buffer, how about we don't do it for shared buffer case in unbind function ?

Reply:
@RanderWang why is this an issue? What does it break? Running on a single core or on different cores?